execution-engine/src/wasi/strace.rs

```diff
@@ -0,0 +1,274 @@
+//! information on licensing and copyright.
+
+use super::common::MemoryHandler;
+use crate::fs::FileSystemResult;
+use std::fmt;
+
+enum TraceState {
+    Func,
+    Args,
+    Result,
+}
+
+pub struct Strace {
+    enabled: bool,
+    state: TraceState,
+}
+
+// Convert a vector of bytes into a printable ASCII string.
+// The string may contain readable text, which is useful to display,
+// but it may also contain binary data, so we cannot decode it as UTF-8.
+fn strace_string(bytes: &[u8], max: usize) -> String {
+    let mut res = String::from("\"");
+    let n = if bytes.len() > max { max } else { bytes.len() };
+    for i in 0..n {
+        if 0x20 <= bytes[i] && bytes[i] < 0x7f {
+            if bytes[i] == b'\\' || bytes[i] == b'"' {
+                res.push_str("\\");
+            }
+            res.push_str(&String::from_utf8_lossy(&bytes[i..i + 1].to_vec()))
+        } else if bytes[i] == 9 {
+            res.push_str("\\t")
+        } else if bytes[i] == 10 {
+            res.push_str("\\n")
+        } else if i + 1 < n && b'0' <= bytes[i + 1] && bytes[i + 1] <= b'9' {
+            // The following character is a digit, so use three octal digits.
+            res.push_str(&format!("\\{:03o}", bytes[i]))
+        } else {
+            res.push_str(&format!("\\{:o}", bytes[i]))
+        }
+    }
+    res.push_str("\"");
+    if bytes.len() > max {
+        res.push_str("...")
+    }
+    res
+}
+
+impl Strace {
+    pub fn func(enabled: bool, name: &str) -> Self {
+        if enabled {
+            eprint!("{}(", name)
+        };
+        Strace {
+            enabled,
+            state: TraceState::Func,
+        }
+    }
+
+    fn arg(&mut self) -> bool {
+        if !self.enabled {
+            return true;
+        }
+        match self.state {
+            TraceState::Func => self.state = TraceState::Args,
+            TraceState::Args => {
+                eprint!(", ");
+            }
+            TraceState::Result => eprint!("\nUnexpected strace arg: "),
+        }
+        false
+    }
+
+    pub fn arg_buffer<T: MemoryHandler>(&mut self, mem: &mut T, adr: u32, len: u32) {
+        if self.arg() {
+            return;
+        }
+        let mut bytes = vec![0u8; len as usize];
+        match mem.read_buffer(adr, &mut bytes) {
+            Ok(()) => eprint!("{}", strace_string(&bytes, 32)),
+            Err(_) => eprint!("BAD_MEM_REF"),
+        }
+    }
+
+    pub fn arg_dec<T: fmt::Display>(&mut self, n: T) {
+        if self.arg() {
+            return;
+        }
+        eprint!("{}", n)
+    }
+
+    pub fn arg_dots(&mut self) {
+        if self.arg() {
+            return;
+        }
+        eprint!("...")
+    }
+
+    pub fn arg_dirents<T: MemoryHandler>(
+        &mut self,
+        _mem: &mut T,
+        _buf_ptr: u32,
+        _buf_len: u32,
+        _result_ptr: u32,
+    ) {
+        if self.arg() {
+            return;
+        }
+        eprint!("DIRENTS") //xx
```

Comment:
What should I put instead of `xx`?

Reply:
I'll mark this as resolved, as this thread of conversation is a duplicate of another thread above (re: using the issue tracker).